### ISSUE TITLE
Dependency error on u-root's cmds/core/backoff

### ISSUE THREAD

**Issue by niels-moller** [u-root/gobusybox#104]
I'm attempting to migrate u-root scripts and delete the old GO111MODULE=off (we used to use `GO111MODULE=off go get ...` to get u-root sources), but I'm getting dependency failures. Below is a script that I would expect to work, and which I hope you can use to reproduce the problem:

```
#! /bin/sh

set -e

GOPATH=${HOME}/tmp/go
export GOPATH

if [ -d "${GOPATH}/pkg/mod" ] ; then chmod u+w -R "${GOPATH}/pkg/mod"; fi
rm -rf ${GOPATH}

UROOT_MOD=github.com/u-root/u-root@v0.11.0

go mod download -x ${UROOT_MOD}

UROOT_SRC=$(go env GOMODCACHE)/${UROOT_MOD}
# Attempt to also download u-root deps
(cd ${UROOT_SRC} && go mod download -x)

go run ${UROOT_MOD} -build=bb -uroot-source=${UROOT_SRC} -o ${HOME}/tmp/initrd core
```

When running this, it downloads a bunch of modules, in particular, I do get a directory `${HOME}/tmp/go/pkg/mod/github.com/cenkalti/backoff/v4@v4.1.3/`. But the u-root invocation fails, with this build error:

```
09:47:45 Build environment: GOARCH=amd64 GOOS=linux GOPATH=/home/nisse/tmp/go CGO_ENABLED=0 GO111MODULE= GOROOT=/usr/lib/go-1.19 PATH=/usr/lib/go-1.19/bin:$PATH
09:47:45 NOTE: building with the new gobusybox; to get the old behavior check out commit 8b790de
09:47:45 Skipping package github.com/u-root/u-root/cmds/core/bind because build constraints exclude all Go files
09:47:45 Gobusybox error: finding packages failed: could not load packages from file system: failed to add package github.com/u-root/u-root/cmds/core/backoff for errors: 3 errors occurred:
        * /home/nisse/tmp/go/pkg/mod/github.com/u-root/u-root@v0.11.0/cmds/core/backoff/run.go:12:2: could not import github.com/cenkalti/backoff/v4 (invalid package name: "")
        * /home/nisse/tmp/go/pkg/mod/github.com/u-root/u-root@v0.11.0/cmds/core/backoff/run.go:23:7: undeclared name: backoff
        * /home/nisse/tmp/go/pkg/mod/github.com/u-root/u-root@v0.11.0/cmds/core/backoff/run.go:40:9: undeclared name: backoff

09:47:45 Build error: error building: preserving bb generated source directory at /tmp/u-root3330130979/builder1707272484 due to error:
finding packages failed: could not load packages from file system: failed to add package github.com/u-root/u-root/cmds/core/backoff for errors: 3 errors occurred:
        * /home/nisse/tmp/go/pkg/mod/github.com/u-root/u-root@v0.11.0/cmds/core/backoff/run.go:12:2: could not import github.com/cenkalti/backoff/v4 (invalid package name: "")
        * /home/nisse/tmp/go/pkg/mod/github.com/u-root/u-root@v0.11.0/cmds/core/backoff/run.go:23:7: undeclared name: backoff
        * /home/nisse/tmp/go/pkg/mod/github.com/u-root/u-root@v0.11.0/cmds/core/backoff/run.go:40:9: undeclared name: backoff

exit status 1
```

Of these error messages, I find `invalid package name: ""` particularly confusing, and I guess that's from the go toolchain, not gobusybox or u-root? Also `preserving bb generated source directory at /tmp/u-root3330130979/builder1707272484` seems wrong, there's no such directory after u-root is finished, so I can't easily look at the generated source dir.

I need some advice on what to do, in order to get u-root's build to locate the `github.com/cenkalti/backoff/v4` dependency.

**Comment by hugelgupf** [comment]
Hi,

I reproduced this. I don't think building directly from the source in $(go env GOMODCACHE)/<package> is really a supported action.

Try this:

```sh
$ cd $(go env GOMODCACHE)/github.com/u-root/u-root@v0.11.0/cmds/core/backoff
$ go build
```

What I got is: 

```
run.go:12:2: cannot find package "." in:
        /home/chrisko/go/pkg/mod/github.com/u-root/u-root@v0.11.0/vendor/github.com/cenkalti/backoff/v4
```

```
$ ls -l $(go env GOMODCACHE)/github.com/u-root/u-root@v0.11.0/vendor 
total 12K
-r--r--r-- 1 chrisko chrisko 12K Jan 10 05:51 modules.txt
```

It's not a full checkout of the git repo. I think as of Go 1.19 (?) even if GO111MODULE=on, the default behavior is still to read the source from vendor unless an additional arg of `-mod=readonly` is passed to `go build`.

```sh
$ cd $(go env GOMODCACHE)/github.com/u-root/u-root@v0.11.0/cmds/core/backoff
$ go1.19.8 build -mod=readonly -o $HOME/backoff
$ echo $?
0
```

gobusybox and u-root do pass -mod=readonly underneath **now** when GO111MODULE=on is set and the behavior can be controlled with u-root's `-go-mod=` (values same as for `go build`: `mod`, `vendor`, `readonly`). But that's a very recent addition.

I'll say though, we should really release a new u-root. A lot has happened since v0.11.0.

**Comment by hugelgupf** [comment]
So my recommendation would be: Do a full checkout of the u-root repo (no need to check out the dependencies).

**Comment by niels-moller** [comment]
I'm now doing a full checkout, and that seems to work. Thanks for the advice.

To get the proper version as specified in go.mod, I use `go list -m -f '{{.Version}}' github.com/u-root/u-root` to select which version to checkout. That's mostly reasonable, but I see some drawbacks:

1. I don't get any verification that the checked out source code matches the corresponding checksum in go.sum.
2. Supports plain version tags, but probably not other kinds of versions that may be present in go.mod.
3. I don't think it will not do the right thing with go.work in all cases. At least, it will not respect a go.work that attempts to point out a local u-root directory, possibly with local changes.

If I could use sources under $(go end GOMODCACHE), I think that would at least solve issue (1).

**Comment by hugelgupf** [comment]
Yeah, I have to admit we haven't found a good way to do this internally either. Generally speaking, gobusybox is trying to support building a Go command whenever it's also buildable in standard Go, and that doesn't seem to be the case here.

(I do get the issue though -- trying to build a command from only local sources without going crazy trying to maintain 20 different local checkouts. We haven't found a good solution for this in our internal builds either, as they can't depend on "online" sources.)

**Comment by hugelgupf** [comment]
Hey, I released a v0.2.0 that changed a lot about how module dependencies work. My recommendation would be to use something like https://github.com/u-root/gobusybox?tab=readme-ov-file#makebb-with-multiple-go-modules -- create a go.mod with u-root in it, and build from that.
